File: src/test/cpp/datastructures/trees/binarysearchtrees/avltree/avl_tree_test.cpp
#include <gtest/gtest.h>

#include <cmath>
#include <vector>
#include <set>
#include <unordered_set>

#include "main/cpp/datastructures/trees/binarysearchtrees/avltree/AVLTree.h"

#include "main/cpp/my_utils.h"

TEST(BinarySearchTreesTest, AVLTreeTest) {
  std::vector<int> values = my_utils::createRandVector(50, 0, 200);
  std::set<int> valueSet;
  for (int i = 0; i < values.size(); i++)
    valueSet.insert(values[i]);

  AVLTree<int> tree;
  for (int i = 0; i < values.size(); i++) {
    tree.insert(values[i]);
    ASSERT_TRUE(tree.isValid());
    ASSERT_LE(tree.getHeight(), 2*std::log2(tree.numNodes()+1));
  }
  ASSERT_EQ(tree.numNodes(), valueSet.size());
  for (int i = 0; i < values.size(); i++) {
    ASSERT_TRUE(tree.contains(values[i]));
  }
  for (int i = 0; i <= 200; i++) {
    ASSERT_EQ(tree.contains(i), valueSet.count(i));
  }
  for (int i = 0; i < values.size(); i++) {
    tree.remove(values[i]);
    ASSERT_TRUE(tree.isValid());
  }
  ASSERT_EQ(tree.numNodes(), 0);
}

File: java/src/test/java/DataStructures/Trees/BinarySearchTrees/AVLTreeTest.java
package DataStructures.Trees.BinarySearchTrees;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import datastructures.trees.binarysearchtrees.avltree.AVLTree;
import org.junit.Before;
import org.junit.Test;

public class AVLTreeTest {

  private AVLTree<Integer> tree;

  @Before
  public void setup() {
    tree = new AVLTree<>();
  }

  @Test
  public void insertTest() {
    assertFalse(tree.contains(5));
    tree.insert(5);
    assertTrue(tree.contains(5));

    for (int i = 0; i < 100; i++) {
      tree.insert((int) (Math.random() * 100));
      assertTrue(tree.isValid());
    }
  }

  @Test
  public void deleteTest() {
    tree.insert(5);
    assertTrue(tree.contains(5));
    tree.delete(5);
    assertFalse(tree.contains(5));

    int[] values = new int[100];

    for (int i = 0; i < values.length; i++) {
      values[i] = (int) (Math.random() * 100);
      tree.insert(values[i]);
    }

    for (int i : values) {
      tree.delete(i);
      assertTrue(tree.isValid());
    }
  }
}
